core/src/main/java/org/apache/iceberg/io/datafile/DataFileServiceRegistry.java

```diff
@@ -0,0 +1,420 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io.datafile;
+
+import java.util.Map;
+import java.util.ServiceLoader;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.base.Objects;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Registry which maintains the available {@link ReaderService} and {@link WriterService}
+ * implementations. Based on the `file format`, the required `data type` and the reader/writer
+ * `builderType` the registry returns the correct reader and writer service implementations. These
+ * services could be used to generate the correct reader and writer builders.
+ */
+public final class DataFileServiceRegistry {
```

Comment:
I commented elsewhere but will add here as well.  I don't like the idea of loading file formats like this because it allows for loading formats that are not officially supported.  I think we need to be careful about what we formally adopt so that we don't end up with data that's not interoperable.  We already have this problem with ORC and to some extent AVRO.

Also SPI is not something we use anywhere else, so I would suggest an approach that follows what was done with InternalData to load supported formats (we had this same discussion in that PR).

Reply:
Move all of the SPI related discussion to a single thread: https://github.com/apache/iceberg/pull/12298#discussion_r1966319213